File: app/macos/NativeStream/NativeStream/Services/ControlSession.swift
//
//  ControlSession.swift
//  NativeStream
//
//  Created by Fredrick Karuri on 30/06/2026.
//


/// Services/ControlSession.swift
///
/// WebSocket client for Local Media Connect control plane.
/// Connects to ws://server/ws, registers as target, and exposes
/// incoming messages via AsyncStream. Reconnects automatically on
/// unexpected close with exponential backoff.

import Foundation
import Observation

private let pingIntervalSeconds: TimeInterval     = 30
private let reconnectBaseDelaySeconds: TimeInterval = 1
private let reconnectMaxDelaySeconds: TimeInterval  = 30

@Observable
@MainActor
final class ControlSession {

    var connected: Bool = false

    private var task: URLSessionWebSocketTask?
    private var session = URLSession(configuration: .default)
    
    @ObservationIgnored
    private var continuation: AsyncStream<Envelope>.Continuation?

    @ObservationIgnored
    let incomingMessages: AsyncStream<Envelope>

    private var explicitDisconnect = false
    private var reconnectDelay: TimeInterval = reconnectBaseDelaySeconds
    private var lastServerURL: URL?
    private var lastDeviceID = ""
    private var lastDeviceName = ""
    
    init() {
        var cont: AsyncStream<Envelope>.Continuation!
        self.incomingMessages = AsyncStream { continuation in
            cont = continuation
        }
        self.continuation = cont
    }

    func connect(serverURL: URL, deviceID: String, deviceName: String) async {
        explicitDisconnect = false
        lastServerURL  = serverURL
        lastDeviceID   = deviceID
        lastDeviceName = deviceName

        guard let wsURL = makeWebSocketURL(from: serverURL) else { return }

        let newTask = session.webSocketTask(with: wsURL)
        task = newTask
        newTask.resume()

        await register(deviceID: deviceID, name: deviceName)
        connected = true
        receive()
    }

    func send(_ envelope: Envelope) async {
        guard let data = try? JSONEncoder().encode(envelope),
              let text = String(data: data, encoding: .utf8) else { return }
        try? await task?.send(.string(text))
    }

    func disconnect() {
        explicitDisconnect = true
        task?.cancel(with: .normalClosure, reason: nil)
        task = nil
        connected = false
    }

    // MARK: - Private

    private func register(deviceID: String, name: String) async {
        guard let envelope = Envelope.encoding(
            type: .register, from: deviceID, to: "server",
            payload: RegisterPayload(name: name, kind: .target)
        ) else { return }
        await send(envelope)
    }

    private func receive() {
        task?.receive { [weak self] result in
            guard let self else { return }
            Task { @MainActor in
                switch result {
                case .success(let message):
                    self.handleMessage(message)
                    self.receive() // continue loop
                case .failure(let error):
                    self.connected = false
                    print("[ControlSession] receive failed: \(error)")
                    if !self.explicitDisconnect { self.scheduleReconnect() }
                }
            }
        }
    }

    private func handleMessage(_ message: URLSessionWebSocketTask.Message) {
        guard case .string(let text) = message,
              let data = text.data(using: .utf8),
              let envelope = try? JSONDecoder().decode(Envelope.self, from: data)
        else { return }
        continuation?.yield(envelope)
    }

    private func scheduleReconnect() {
        guard let serverURL = lastServerURL else { return }
        Task {
            try? await Task.sleep(for: .seconds(reconnectDelay))
            reconnectDelay = min(reconnectDelay * 2, reconnectMaxDelaySeconds)
            await connect(serverURL: serverURL, deviceID: lastDeviceID, deviceName: lastDeviceName)
        }
    }

    private func makeWebSocketURL(from httpURL: URL) -> URL? {
        var components = URLComponents(url: httpURL, resolvingAgainstBaseURL: false)
        components?.scheme = "ws"
        components?.path = "/ws"
        return components?.url
    }
}

File: app/android/app/src/main/java/com/nativestream/android/data/remote/ControlSession.kt
// app/src/main/java/com/nativestream/android/data/remote/ControlSession.kt
//
// WebSocket client for Local Media Connect control plane.
// Connects to ws://server/ws, registers as controller, and exposes
// inbound envelopes as a SharedFlow for ControlViewModel to consume.
// Reconnects automatically on unexpected close with exponential backoff.

package com.nativestream.android.data.remote

import android.util.Log
import com.nativestream.android.data.local.SettingsDataStore
import com.nativestream.android.domain.model.control.DeviceKind
import com.nativestream.android.domain.model.control.Envelope
import com.nativestream.android.domain.model.control.MessageType
import com.nativestream.android.domain.model.control.RegisterPayload
import com.nativestream.android.domain.model.control.buildEnvelope
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

private const val TAG                     = "ControlSession"
private const val PING_INTERVAL_SECONDS   = 30L
private const val RECONNECT_BASE_DELAY_MS = 1_000L
private const val RECONNECT_MAX_DELAY_MS  = 30_000L
private const val NORMAL_CLOSE_CODE       = 1000

@Singleton
class ControlSession @Inject constructor(
    private val settingsDataStore: SettingsDataStore,
) {
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private val _messages = MutableSharedFlow<Envelope>(extraBufferCapacity = 64)
    val messages: SharedFlow<Envelope> = _messages

    private val _connected = MutableStateFlow(false)
    val connected: StateFlow<Boolean> = _connected

    private var webSocket: WebSocket? = null
    private var explicitDisconnect = false
    private var reconnectDelayMs   = RECONNECT_BASE_DELAY_MS

    private val client = OkHttpClient.Builder()
        .pingInterval(PING_INTERVAL_SECONDS, TimeUnit.SECONDS)
        .build()

    private val json = Json { ignoreUnknownKeys = true }

    // ── Public API ────────────────────────────────────────────────────────────

    fun connect(serverUrl: String, deviceId: String, deviceName: String) {
        explicitDisconnect = false
        val wsUrl = serverUrl
            .removePrefix("http://")
            .removePrefix("https://")
            .let { "ws://$it/ws" }

        Log.d(TAG, "Connecting to $wsUrl")

        val request = Request.Builder().url(wsUrl).build()
        webSocket = client.newWebSocket(request, object : WebSocketListener() {

            override fun onOpen(ws: WebSocket, response: Response) {
                _connected.value  = true
                reconnectDelayMs  = RECONNECT_BASE_DELAY_MS
                Log.d(TAG, "WebSocket connected")

                // Register as controller immediately on open
                val envelope = buildEnvelope(
                    type    = MessageType.REGISTER,
                    from    = deviceId,
                    to      = "server",
                    payload = RegisterPayload(name = deviceName, kind = DeviceKind.CONTROLLER),
                )
                ws.send(json.encodeToString(envelope))
            }

            override fun onMessage(ws: WebSocket, text: String) {
                runCatching {
                    val envelope = json.decodeFromString<Envelope>(text)
                    scope.launch { _messages.emit(envelope) }
                }.onFailure {
                    Log.w(TAG, "Failed to parse envelope: $text", it)
                }
            }

            override fun onClosing(ws: WebSocket, code: Int, reason: String) {
                ws.close(NORMAL_CLOSE_CODE, null)
            }

            override fun onClosed(ws: WebSocket, code: Int, reason: String) {
                _connected.value = false
                Log.d(TAG, "WebSocket closed: $code $reason")
                if (!explicitDisconnect) scheduleReconnect(serverUrl, deviceId, deviceName)
            }

            override fun onFailure(ws: WebSocket, t: Throwable, response: Response?) {
                _connected.value = false
                Log.w(TAG, "WebSocket failure", t)
                if (!explicitDisconnect) scheduleReconnect(serverUrl, deviceId, deviceName)
            }
        })
    }

    fun send(envelope: Envelope) {
        val text = runCatching { json.encodeToString(envelope) }.getOrNull() ?: return
        webSocket?.send(text)
    }

    fun disconnect() {
        explicitDisconnect = true
        webSocket?.close(NORMAL_CLOSE_CODE, "explicit disconnect")
        webSocket      = null
        _connected.value = false
    }

    // ── Reconnect ─────────────────────────────────────────────────────────────

    private fun scheduleReconnect(serverUrl: String, deviceId: String, deviceName: String) {
        scope.launch {
            Log.d(TAG, "Reconnecting in ${reconnectDelayMs}ms")
            delay(reconnectDelayMs)
            reconnectDelayMs = minOf(reconnectDelayMs * 2, RECONNECT_MAX_DELAY_MS)
            connect(serverUrl, deviceId, deviceName)
        }
    }
}